util/stop: allow handling of panics

tbg opened the pull request:
The Stopper is our main source of goroutines, which means that many
panics will pass through it, providing a good opportunity for insertion
of a reporting hook at some point in the future.

This is provided now since we've recently looked at some error reporting
tools, and this will come in handy.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7764)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[util/stop/stopper.go, line 136 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764#-KMU50DCcADvCJ5al0xr:-KMU50DCcADvCJ5al0xs:-648071948) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bc3bb38a0798c2ee8984d8e59fbecb161fda8d/util/stop/stopper.go#L136)):*
> ```Go
> 
> // OnPanic is an option which lets the Stopper recover from all panics using
> // the provided panic handler.
> ```

Add a blank line between paragraphs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764#-:-KMU5ErAERTyiPzfjPAz:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
PS The PR title should read `util/stop: allow handling of panics`.

tbg committed:
util/stop: allow handling of panics

The Stopper is our main source of goroutines, which means that many
panics will pass through it, providing a good opportunity for insertion
of a reporting hook at some point in the future.

tschottdorf commented:
I don't think putting the full path here is a good idea (I'll do it here anyway because it doesn't matter for this one).
The subject line should usually not exceed 50 characters, and while it's usually doable to give a concise summary with that limitation, it gets harder the more cruft we add to the beginning.

Think about where this goes: `storage/storagebase:` is already 20 characters. While I think that some idea about which part of the system was touched is useful, anything that requires us to put full paths there should be moved into a commit marker.

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[util/stop/stopper.go, line 136 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764#-KMU50DCcADvCJ5al0xr:-KMUGDqUthIUjEixdj_T:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bc3bb38a0798c2ee8984d8e59fbecb161fda8d/util/stop/stopper.go#L136)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Add a blank line between paragraphs.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Putting the full path is the intention. I always see 50 characters as the recommendation, not a firm limit.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[util/stop/stopper.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764#-KMUKUHoUAkcdhSLviP5:-KMUKUHoUAkcdhSLviP6:1529161551) ([raw file](https://github.com/cockroachdb/cockroach/blob/e68df029829e8022683526de8001ba033cccc450/util/stop/stopper.go#L164)):*
> ```Go
> func (s *Stopper) maybeHandlePanic() {
> 	if r := recover(); r != nil {
> 		if s.onPanic != nil {
> ```

don't you want to check this first to avoid an unnecessary adjustment to the stack trace?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764#-:-KMUKglUKFnEnJN89fwa:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[util/stop/stopper.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764#-KMUKUHoUAkcdhSLviP5:-KMULGoJwdDiWGIqIG1O:-1978676072) ([raw file](https://github.com/cockroachdb/cockroach/blob/e68df029829e8022683526de8001ba033cccc450/util/stop/stopper.go#L164)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> don't you want to check this first to avoid an unnecessary adjustment to the stack trace?
</details>
Ah, just merged. I'll keep this in mind for future drive-by.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7764)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Go package naming conventions mean that the last component of our path names are generally unique, so I don't think we need the full path in the commit message. 

We also don't enforce the 50-character rule for the first line of commit messages. I can't find the thread but we decided that since we wrap code at 100 columns instead of 80 we should allow 70 characters in commit messages (adding 20 columns each time).

tschottdorf commented:
We don't enforce either of them (well, Pete seems to enforce one of them).
I'd much rather just do what is suggested to every person using Git instead
of cooking our own here, but I don't really care as long as it doesn't get
out of hand.

On Tue, Jul 12, 2016 at 10:16 PM Ben Darnell <notifications@github.com>
wrote:

> Go package naming conventions mean that the last component of our path
> names are generally unique, so I don't think we need the full path in the
> commit message.
>
> We also don't enforce the 50-character rule for the first line of commit
> messages. I can't find the thread but we decided that since we wrap code at
> 100 columns instead of 80 we should allow 70 characters in commit messages
> (adding 20 columns each time).
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7764#issuecomment-232235925>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135IsrusuutBRVCKNIzTothgT98SYDks5qVEqKgaJpZM4JKRNk>
> .
>
-- 

-- Tobias

petermattis commented:
The prefix before the `:` is used in the [release dashboard](https://s3.amazonaws.com/cockroachdb/issues/beta.html) for grouping. For issues, the unwritten convention was to use the primary package that the issue affected. I have an unexecuted plan to incorporate PRs into that dashboard. Note that this is following the convention used for the [Go release dashboard](https://dev.golang.org/release).

We're not locked into following what the Go-team does here. Happy to discuss this on gitter or elsewhere.
